bevy_lint/src/lints/zst_query.rs

```diff
@@ -0,0 +1,141 @@
+//! Checks for queries that query for a zero-sized type.
+//!
+//! # Motivation
+//!
+//! Zero-sized types (ZSTs) are types that have no size as a result of containing no runtime data.
+//! In Bevy, such types are often used as marker components and are best used as filters.
+//!
+//! # Example
+//!
+//! ```
+//! # use bevy::prelude::*;
+//!
+//! #[derive(Component)]
+//! struct Player;
+//!
+//! fn move_player(mut query: Query<(&mut Transform, &Player)>) {
+//!     // ...
+//! }
+//! ```
+//!
+//! Use instead:
+//!
+//! ```
+//! # use bevy::prelude::*;
+//!
+//! #[derive(Component)]
+//! struct Player;
+//!
+//! fn move_player(query: Query<&mut Transform, With<Player>>) {
+//!     // ...
+//! }
+//! ```
+
+use crate::{
+    declare_bevy_lint,
+    utils::hir_parse::{detuple, generic_type_at},
+};
+use clippy_utils::{
+    diagnostics::span_lint_and_help,
+    ty::{is_normalizable, match_type},
+};
+use rustc_abi::Size;
+use rustc_hir_analysis::collect::ItemCtxt;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::{
+    layout::{LayoutOf, TyAndLayout},
+    Ty,
+};
+use rustc_session::declare_lint_pass;
+
+declare_bevy_lint! {
+    pub ZST_QUERY,
+    RESTRICTION,
+    "query for a zero-sized type"
+}
+
+declare_lint_pass! {
+    ZstQuery => [ZST_QUERY.lint]
+}
+
+impl<'tcx> LateLintPass<'tcx> for ZstQuery {
+    fn check_ty(&mut self, cx: &LateContext<'tcx>, hir_ty: &'tcx rustc_hir::Ty<'tcx>) {
+        let item_cx = ItemCtxt::new(cx.tcx, hir_ty.hir_id.owner.def_id);
+        let ty = item_cx.lower_ty(hir_ty);
+
+        let Some(query_kind) = QueryKind::try_from_ty(cx, ty) else {
+            return;
+        };
+
+        let Some(query_data_ty) = generic_type_at(cx, hir_ty, 2) else {
+            return;
+        };
+
+        for hir_ty in detuple(*query_data_ty) {
```

Comment:
```suggestion
        for hir_ty in detuple(*query_data_ty.peel_refs()) {
```

If you remove the `peel_refs()` for `detuple()` from a previous comment, you'll need to add it back here.